src/metabase/api/api_key.clj

```diff
@@ -29,28 +30,84 @@
+        email        (format "api-key-user-%s@api-key.invalid" (u/slugify name))]
     (t2/with-transaction [_conn]
       (let [user (user/insert-new-user! {:email      email
                                          :first_name name
                                          :type       :api-key})]
         (user/set-permissions-groups! user [(perms-group/all-users) group_id])
-        (-> (t2/insert-returning-instance! :model/ApiKey
-                                           {:user_id      (u/the-id user)
-                                            :name         name
-                                            :unhashed_key api-key
-                                            :created_by   api/*current-user-id*})
-            (select-keys [:created_at :updated_at :id])
-            (assoc :name name
-                   :group_id group_id
-                   :unmasked_key api-key
-                   :masked_key (api-key/mask api-key)))))))
+        (let [api-key (-> (t2/insert-returning-instance! :model/ApiKey
+                                                         {:user_id      (u/the-id user)
+                                                          :name         name
+                                                          :unhashed_key unhashed-key
+                                                          :created_by   api/*current-user-id*})
+                          (t2/hydrate :group_name))]
+          (events/publish-event! :event/api-key-create
+                                 {:object  api-key
+                                  :user-id api/*current-user-id*})
+          (-> api-key
+              (select-keys [:created_at :updated_at :id :group_name :name])
+              (assoc :unmasked_key unhashed-key
+                     :masked_key (api-key/mask unhashed-key))))))))
 
 (api/defendpoint GET "/count"
   "Get the count of API keys in the DB"
   [:as _body]
   (api/check-superuser)
   (t2/count :model/ApiKey))
 
+(api/defendpoint PUT "/:id"
+  "Update an API key by changing its group and/or its name"
+  [id :as {{:keys [group_id name] :as _body} :body}]
+  {id       ms/PositiveInt
+   group_id [:maybe ms/PositiveInt]
+   name     [:maybe ms/NonBlankString]}
+  (api/check-superuser)
+  (t2/with-transaction [_conn]
+    (let [api-key-before (-> (t2/select-one :model/ApiKey :id id)
+                             ;; hydrate the group_name for audit logging
+                             (t2/hydrate :group_name))]
+      (when name
+        ;; A bit of a pain to keep these in sync, but oh well.
+        (t2/update! :model/User (:user_id api-key-before) {:first_name name})
+        (t2/update! :model/ApiKey id {:name name}))
+      (when group_id
+        (let [user (-> api-key-before (t2/hydrate :user) :user)]
+          (user/set-permissions-groups! user [(perms-group/all-users) {:id group_id}])))
+      (let [updated-api-key (-> (t2/select-one :model/ApiKey :id id)
+                                (t2/hydrate :group_name))]
```

Comment:
nit: Is it possible to create `updated-api-key` here by just updating the relevant fields on `api-key-before` instead of hitting the DB again? Similar to what you're doing below with the regenerate endpoint.

Reply:
I should maybe revisit this because I think it's possibly a sign that the design could use improving, but the problem is the hydration method for `group_name`. We hydrate the group name earlier (to get the previous name for audit logging) and now need to do it again - but calling `t2/hydrate!` a second time won't do anything if we've already hydrated before.

I'd definitely be happy to pull something out of the hydration function to allow us to go from a `group_id` to a `group_name` more easily, and reuse that here. Would you be ok with me doing that in a future PR?

Reply:
Gotcha, yeah I was just curious if that was possible or not. No need for it to block the PR